### PR TITLE
feat: periodically send did-you-know messages to users

### PR DESCRIPTION
**Pull request by mamuqsit**
## Introduction
This feature sends ASHAbot users factual reminders, i.e., "Did you know?" (DYK) messages periodically - roughly on a biweekly basis, and in their preferred language. Facts are pulled from [`dyk_v1.csv`](https://github.com/A4i-tech/byoeb/blob/b4fb897469ada0d668ef69991f9e2a4bec6b5d69/byoeb-v1/byoeb/byoeb/background_jobs/did_you_know/dyk.csv).

## Candidate selection
A valid candidate user is one having a supported DYK language (currently, Hindi and English) and has no pending DYKs scheduled. With `APP_ENV=PROD`, only users with the "asha" role are selected. Otherwise, only test users (`test_user=True`) are picked.

Ofcourse, we are working with a rather small (and definitely exhaustive) dataset of facts. Users who have been sent all possible DYK messages (i.e., reached "exhaustion") will not be selected either.

## Logging
There are 2 loggers. "dyk_run" logger sends script errors (like file-not-found) and aggregate statistics. "dyk_send" logger logs every queued and dispatched DYK messages (including failed tasks and their reason for failure).
```log
2025-10-15 20:06:43,573 dyk_run INFO: Synced jobs: 0
2025-10-15 20:06:43,577 dyk_run INFO: Queued jobs: 0
2025-10-15 20:06:43,578 dyk_run INFO: Exhausted jobs: 0
2025-10-15 20:06:44,165 dyk_send INFO: Sent DYK 7b0785b4-a6b2-4dc3-9786-d53b438cdff8 to user b63b2c24f5024230d6fa624712980b80
2025-10-15 20:06:44,170 dyk_run INFO: Dispatched jobs: 1 succeeded, 0 failed
```
A "mirroring" strategy is used to mirror these logs to Azure AppInsights when `APPINSIGHTS_CONNECTION_STRING` environment variable is provided to the script. Structured logging is used here to make these messages query-friendly, so you can actually lookup DYK logs for specific user IDs and phone numbers using KQL:
```kql
customEvents
| where name == "dyk_send"
| where customDimensions.["details.user_phone_number"] == "91xxxxxxxxxx"
```
image

The following events are logged with their log message and log level (info / warning / error) in `customEvents` table:
- Number of pending messages discarded during sync
- Number of messages added to the dispatch queue
- Number of users who have exhausted all DYK messages
- Number of messages successfully dispatched to WhatsApp
- Number of messages that failed during dispatch
- Details of each successfully sent DYK message (customDimensions props: `dyk_id` `user_id` `user_phone_number` `whatsapp_message_ids`)
- Details of each failed dispatch DYK message (customDimensions props: `dyk_id` `user_id` `user_phone_number`)
- User exhaustion events during queueing (customDimensions props: `user_id` `user_phone_number`)
- Any retry attempts during dispatch
- Missing file or configuration errors at startup

## Contributing facts to the dataset
The DYK CSV file has 3 columns:
| en                                                   | hi                                                  | GUID                                   |
|------------------------------------------------------------------|------------------------------------------------------------------------|-----------------------------------------|
| Drinking alcohol during pregnancy can lead to long-term harm to the baby. | गर्भावस्था के दौरान शराब पीने से बच्चे को दीर्घकालिक नुकसान हो सकता है। | 8e1c55a7-85d6-44ec-8efd-973fc3296b1f |

To add a new fact, simply add a row with a randomly generated GUID. Likewise, to remove an existing fact, remove the complete row. Empty strings under language cols ("en", "hi") are valid values, in which case these DYKs will not be selected for users of that language.

To support a new language, add a new column named by the language's code, e.g. `mr` (ensure it is registered in [`LanguageCode`](https://github.com/A4i-tech/byoeb/blob/f8abacc4d4511bff766f229a4d52a92db311bad2/byoeb-v1/byoeb/byoeb/constants/user_enums.py)). Then, append the "language" field in `background_jobs/did_you_know/bot_config.json` with:
```jsonc
{
    "language": "mr",
    "template": [
        "💡 *तुम्हाला माहित आहे का??*",  // <-- "Did you know" translated in Marathi
        "{message}"
    ]
}
```

To test the DYK feature, execute `send_dyk.py` with the env vars listed below (set values accordingly).
```sh
$ cd byoeb-v1/byoeb
$ APP_ENV=TEST \
  APPINSIGHTS_CONNECTION_STRING="" \
  WHATSAPP_VERIFICATION_TOKEN=98765  \
  WHATSAPP_AUTH_TOKEN="" \
  WHATSAPP_PHONE_NUMBER_ID="" \
  poetry run python3 byoeb/background_jobs/did_you_know/send_dyk.py
```

## Attachments
Preview of WhatsApp messages dispatched by the send_dyk.py background task:
image

### CODE REVIEW

**Comment by mamuqsit** [comment]
## Steps to test
Register yourself as a test user:
1. Run ASHAbot chat app.
2. Register yourself as a test user:
   ```sh
   curl -X POST http://127.0.0.1:5000/register_users \
     -H "Content-Type: application/json" \
     -d '[{"user_id": "", "phone_number_id": "91xxxxxxxxxx", "user_name": "Name User", "user_type": "asha", "user_language": "en", "test_user": true}]'
   # or if you are already registered:
   curl -X POST http://127.0.0.1:5000/update_users \
     -H "Content-Type: application/json" \
     -d '[{"phone_number_id": "91xxxxxxxxxx", "test_user": true}]'
   ```

### A. Test successful dispatches
In the event of one-shot successful dispatches, messages are delivered in one go.
1. Run send_dyk.py by running the command in **Developer notes**
2. You should receive a DYK message on WhatsApp.
3. Look for the following logs. Ensure a non-zero value for "Queued jobs" and succeeded "Dispatched jobs":
   ```sh
   2025-10-16 13:21:26,177 dyk_run INFO: Queued jobs: 1
   2025-10-16 13:21:27,852 dyk_run INFO: Dispatched jobs: 1 succeeded, 0 failed
   ```

### B. Test unsuccessful dispatches
Queue op assigns DYKs to users early during the script. Any failure in dispatch makes next queue ops ignore users that were already assigned a DYK message.
1. Set `WHATSAPP_AUTH_TOKEN` to some invalid token
2. Run send_dyk.py by running the command in **Developer notes**
3. The script should terminate after 5 retry attempts.
4. Look for the following logs. Ensure a non-zero value for "Queued jobs".
   ```sh
   2025-10-16 13:23:31,026 dyk_run INFO: Synced jobs: 0
   2025-10-16 13:23:31,032 dyk_run INFO: Queued jobs: 1
   2025-10-16 13:23:31,033 dyk_run INFO: Exhausted jobs: 0
   2025-10-16 13:23:31,274 dyk_run INFO: Dispatched jobs: 0 succeeded, 1 failed
   ...
   2025-10-16 13:23:41,920 dyk_run WARNING: Retrying dispatch job... 5 / 5
   2025-10-16 13:23:42,134 dyk_run INFO: Dispatched jobs: 0 succeeded, 1 failed
   2025-10-16 13:23:42,135 dyk_run ERROR: Max retries exceeded. Exiting.
   ```
5. Rerun send_dyk.py again (with the invalid WHATSAPP_AUTH_TOKEN)
6. Check logs again - this time, ensure a **zero** value for "Queued jobs"
7. Rerun send_dyk.py again, but with a valid WHATSAPP_AUTH_TOKEN
8. Check logs again - ensure a zero value for "Queued jobs", but a non-zero value for successful "Dispatched jobs".

### C. Test language selection
User must receive DYKs in their preferred language.
1. Change your language to Hindi if your language is set to English, or vice versa:
   ```sh
   curl -X POST http://127.0.0.1:5000/update_users \
     -H "Content-Type: application/json" \
     -d '[{"phone_number_id": "91xxxxxxxxxx", "user_language": "hi"}]'
   ```
2. Run send_dyk.py by running the command in **Developer notes**
3. You should receive a DYK message on WhatsApp in your preferred language

### D. Test Azure AppInsights logging
Running the above tests with `APPINSIGHTS_CONNECTION_STRING` should emit the logs on Azure.
1. Visit https://portal.azure.com and select the Application Insights service
2. In Monitoring → Logs, select `customEvents` table
3. Narrow your search - set `Time range: Last 30 minutes`
4. Narrow your search further - click "Add" and select `name`. Tick `dyk_run` and `dyk_send`
5. Console logs should show up here as well. Note that the logs on Azure lag by 2 minutes (they do not appear in realtime).

**Comment by mamuqsit** [comment]
There is a failure in staging (`ModuleNotFoundError: No module named 'byoeb'`) across scripts in background_jobs. A `poetry add` [here](https://github.com/A4i-tech/byoeb/blob/a97696d86e9f8da67646198cfe2f0d584fe4f1d0/byoeb-v1/byoeb/byoeb/apis/background_jobs.py#38) would be a quick-fix, but there is no guarantee this would work in prod:
```diff
-   f"0 12 * * FRI exec python3 {jobs_path}/message_leaderboard/leaderboard.py; exit",
+   f"0 12 * * FRI exec poetry run python3 {jobs_path}/message_leaderboard/leaderboard.py; exit",
```

Because staging and prod run commands differ (`poetry run python3 byoeb/chat_app/run.py` vs. `byoeb-v1/start.sh`).

However, docker containerization (the CD task) will inadvertently address the difference as a unified `RUN` command in Dockerfile would be used to spin up ASHAbot. Investigating alternatives in the meantime.

File: byoeb-v1/byoeb/byoeb/background_jobs/did_you_know/send_dyk.py [hunk 1]
```diff
@@ -0,0 +1,344 @@
...
+run_logger.setLevel(logging.DEBUG)
+run_logger.addHandler(handler)
+run_logger.addHandler(app_insights_handler)
+
+send_logger = logging.getLogger("dyk_send")
+send_logger.setLevel(logging.DEBUG)
+send_logger.addHandler(handler)
+send_logger.addHandler(app_insights_handler)
+
+current_dir = os.path.dirname(os.path.abspath(__file__))
```
Comment (Soumabha) [inline byoeb-v1/byoeb/byoeb/background_jobs/did_you_know/send_dyk.py]: Move these modifications to the config.py similar to how it is done for other jobs. This would help keep the related file hierarchies aligned.

File: byoeb-v1/byoeb/byoeb/background_jobs/did_you_know/send_dyk.py [hunk 2]
```diff
@@ -0,0 +1,344 @@
...
+    run_logger.error("File no found: %s", SOURCE_PATH)
+    exit(1)
+
+# parse and index facts sheet for quick lookup
+with open(SOURCE_PATH) as f:
+    reader = csv.reader(f)
+
+    # fail fast - if these expected cols dont exist, python will bail early
+    cols = next(reader)
+    lang_cols = {lang.language: cols.index(lang.column) for lang in LANG_ENTRIES.values()}
```
Comment (Soumabha) [inline byoeb-v1/byoeb/byoeb/background_jobs/did_you_know/send_dyk.py]: The usage of index basis for language interpretation seems challenging as it falls on the human to be aware and make csv changes accordingly. I think it is better if we can move to an exact column name match processing no matter which the order is.

The csv having entire 'Did you know' and 'Did you know - Hindi' does not make as much sense.
My proposal would be to copy and modify the dyk.csv with exact col matches to language. (keep a copy within the byoeb-v1 bg-jobs path).

File: byoeb-v1/byoeb/byoeb/background_jobs/did_you_know/send_dyk.py [hunk 3]
```diff
@@ -0,0 +1,344 @@
...
+    """
+    Does a buffered fetch operation on users collection and get their set of sent DYK ids.
+    Collects only users who meet all the following criteria:
+    - the user is an asha user (prod only)
+    - the user is a test user (staging only)
+    - the user has a language available in `langs`
+    - the user has no 'pending' DYKs
+    """
+    match_stage = {"User.user_language": {"$in": [x.value for x in langs]}}
+    if os.getenv("APP_ENV") == "PROD":
```
Comment (Soumabha) [inline byoeb-v1/byoeb/byoeb/background_jobs/did_you_know/send_dyk.py]: Could we make it configuration driven so that we dont need to hardcode on selection details?? 
We can do something like pick test users when the "user_types_to_send" is empty; no matter which environment.

File: byoeb-v1/byoeb/byoeb/background_jobs/did_you_know/send_dyk.py [hunk 4]
```diff
@@ -0,0 +1,344 @@
...
+        match_stage["User.user_type"] = "asha"
+    else:
+        match_stage["User.test_user"] = True
+
+    user_client = await user_db_service._get_collection_client(user_db_service.collection_name)
```
Comment (Soumabha) [inline byoeb-v1/byoeb/byoeb/background_jobs/did_you_know/send_dyk.py]: 1. Can we use the userRepository (services>database>mongo_db> for any interactions with the user collection? That would help us support the MVC pattern of interactions in bg_jobs too where we rely on service functionalities as needed.

2. Also, as the user base grows the query will not scale as we are fetching a huge number; lets implement batching with sufficient batch size to make the design futureproof. I would also suggest doing a batch for the actual workflow rather than individual queries / functionalities.

File: byoeb-v1/byoeb/byoeb/background_jobs/did_you_know/send_dyk.py [hunk 5]
```diff
@@ -0,0 +1,344 @@
...
+    dyk_client = await user_db_service._get_collection_client(dyks_sent_collection_name)
```
Comment (Soumabha) [inline byoeb-v1/byoeb/byoeb/background_jobs/did_you_know/send_dyk.py]: Lets create a repository class for dyk collection too.

File: byoeb-v1/byoeb/byoeb/background_jobs/did_you_know/send_dyk.py [hunk 6]
```diff
@@ -0,0 +1,344 @@
...
+            print("Cannot assign DYK to %s (they received all DYKs)" % phone_number)
+    """
+
+    n_queued = 0
+    n_exhausted = 0
+    lang_sets = {lang: set(messages.keys()) for lang, messages in records.items()}
+    queued_client_ops = []
+    async for user, sent in pick_candidates(records.keys()):
+        lang = LanguageCode(user.user_language)
+        diff = lang_sets[lang] - sent  # deduplication
```
Comment (Soumabha) [inline byoeb-v1/byoeb/byoeb/background_jobs/did_you_know/send_dyk.py:85]: What if this is a Marathi user? We should handle cases where user is onboarded on langX but we do not have dyks for langX as is the current setup for Marathi (unsure about Kannada but that is next).

File: byoeb-v1/byoeb/byoeb/background_jobs/did_you_know/send_dyk.py [hunk 7]
```diff
@@ -0,0 +1,344 @@
...
+        if len(diff) == 0:
+            # no facts remaining !
+            send_logger.warning("User %s is exhausted", user.user_id, extra={"dyk": {
+                "context": queue.__name__,
+                "user_id": user.user_id,
+                "user_phone_number": user.phone_number_id
+            }})
+            continue
+        uuid = random.Random(user.user_id).choice(list(sorted(diff)))
```
Comment (Soumabha) [inline byoeb-v1/byoeb/byoeb/background_jobs/did_you_know/send_dyk.py]: It's not cryptographically secure. The results are predictable if the user ID is known.

Is there a _id equivalent field which acts like the uniqueness in the table? Also since we already have the dyk uuid from the csv, why not just reuse that along with the separate unique id?

File: byoeb-v1/byoeb/byoeb/background_jobs/did_you_know/send_dyk.py [hunk 8]
```diff
@@ -0,0 +1,344 @@
...
+            continue
+
+        phone_number = user_doc["User"]["phone_number_id"]
+        lang = LanguageCode(doc["dyk_lang"])
+        message = LANG_ENTRIES[lang].template.replace("{message}", records[lang][doc["dyk_id"]])
+
+        text_message = ByoebMessageContext(
+            channel_type="whatsapp",
+            message_category="did_you_know",
+            user=User(
```
Comment (Soumabha) [inline byoeb-v1/byoeb/byoeb/background_jobs/did_you_know/send_dyk.py]: We should maintain the same user record as the one fetched, else we wont be able to correlate all the respective messages sent for a particular user.

File: byoeb-v1/byoeb/byoeb/background_jobs/did_you_know/send_dyk.py [hunk 9]
```diff
@@ -0,0 +1,344 @@
...
+                test_user=False,
+                experts={},
+                audience=[],
+                created_timestamp=ts,
+                activity_timestamp=ts,
+                last_conversations=[],
+                additional_info={}
+            ),
+            message_context=MessageContext(
+                message_id=f"did-you-know-{phone_number}-{ts}",
```
Comment (Soumabha) [inline byoeb-v1/byoeb/byoeb/background_jobs/did_you_know/send_dyk.py]: Lets use the unique id from the dyk collection.

**Review by Soumabha — CHANGES_REQUESTED** [review]

File: byoeb-v1/byoeb/byoeb/background_jobs/did_you_know/send_dyk.py (hunk 6 above)
Comment (mamuqsit) [reply byoeb-v1/byoeb/byoeb/background_jobs/did_you_know/send_dyk.py:85]: `pick_candidates` excludes users with unsupported languages:
```py
match_stage = {"User.user_language": {"$in": [x.value for x in langs]}}
```
Marathi users will be left out. Looking back at this, logging aggregate stats may be useful here to know the missed DYK audience.

File: byoeb-v1/byoeb/byoeb/background_jobs/did_you_know/send_dyk.py (hunk 7 above)
Comment (mamuqsit) [reply byoeb-v1/byoeb/byoeb/background_jobs/did_you_know/send_dyk.py]: Agreed this should be changed. However, I don't think we need cryptographically secure randomness here - the performance overhead is not justified when we are just selecting educational messages.

Reproducibility of results by using user_id as seed does not provide meaningful debugging benefits either. Making the generator non-deterministic here should suffice, right?
```diff
- uuid = random.Random(user.user_id).choice(list(sorted(diff)))
+ uuid = random.choice(list(sorted(diff)))
```

File: byoeb-v1/byoeb/byoeb/background_jobs/did_you_know/send_dyk.py (hunk 2 above)
Comment (mamuqsit) [reply byoeb-v1/byoeb/byoeb/background_jobs/did_you_know/send_dyk.py]: Some clarification here, this isn't index-based matching but rather index lookup. A better idea of what's happening here:
```py
lang_cols = {LanguageCode.HINDI: cols.index("Did you know - Hindi")}
```
I've replaced overly human-centric field names - `Did you know - Hindi` with simpler `en` and `hi` labels. Pydantic auto handles col name validation against `LanguageCode` enum.

| en                                                       | hi                                                     | GUID                                   |
|-----------------------------------------------------------|---------------------------------------------------------|-----------------------------------------|
| Vitamin A supplements should not be taken during pregnancy. | गर्भावस्था के दौरान विटामिन ए की खुराक नहीं ली जानी चाहिए। | 2c44ae3b-eb5c-4204-a25d-b20f16d0b3cd |

However, "new language registration" is still a two-step process - need to update dyk.csv _and_ register a message template for it in `bot_config.py`.

File: byoeb-v1/byoeb/byoeb/background_jobs/did_you_know/send_dyk.py (hunk 9 above)
Comment (mamuqsit) [reply byoeb-v1/byoeb/byoeb/background_jobs/did_you_know/send_dyk.py]: Resolved by https://github.com/A4i-tech/byoeb/pull/34/commits/fa48d07372333a02b37339b45e446c4d0d2beeab

File: byoeb-v1/byoeb/byoeb/background_jobs/did_you_know/send_dyk.py (hunk 8 above)
Comment (mamuqsit) [reply byoeb-v1/byoeb/byoeb/background_jobs/did_you_know/send_dyk.py]: Resolved by https://github.com/A4i-tech/byoeb/pull/34/commits/ec50ff8259fa1a5aa568a52e72bd5ae901210429

File: byoeb-v1/byoeb/byoeb/background_jobs/did_you_know/send_dyk.py (hunk 3 above)
Comment (mamuqsit) [reply byoeb-v1/byoeb/byoeb/background_jobs/did_you_know/send_dyk.py]: Addressed by https://github.com/A4i-tech/byoeb/pull/34/commits/b462d7f4b4418e6bae8b8745ba59136a11a3532f
An empty list will select test users with a debug log:
```
2025-10-21 17:12:48,730 dyk_run DEBUG: pick_candidates: user_types list is empty - selecting test users instead
```

File: byoeb-v1/byoeb/byoeb/background_jobs/did_you_know/send_dyk.py (hunk 6 above)
Comment (mamuqsit) [reply byoeb-v1/byoeb/byoeb/background_jobs/did_you_know/send_dyk.py:85]: About unexpected languages, language resolution happens earlier than the `lang = LanguageCode(user.user_language)` here. E.g., registering a language=xx in bot_config.json causes the script to fail early:
```sh
> poetry run python3 byoeb/background_jobs/did_you_know/send_dyk.py
pydantic_core._pydantic_core.ValidationError: 1 validation error for LangEntry
language
  Input should be 'hi', 'en', 'mr' or 'te' [type=enum, input_value='xx', input_type=str]
    For further information visit https://errors.pydantic.dev/2.12/v/enum
```

Similarly, registering 'mr' when there is no 'mr' in dyk.csv fails early as well:
```sh
ValueError: 'mr' is not in list
```

Comment (mamuqsit) [reply byoeb-v1/byoeb/byoeb/background_jobs/did_you_know/send_dyk.py:85]: https://github.com/A4i-tech/byoeb/pull/34/commits/c715a97dd946754c0d46ce569b3990795f08828d elaborates the ValueError above:
```diff
- ValueError: 'mr' is not in list
+ ValueError: Column "mr" does not exist in dyk.csv - did you forget to create a column for "mr"?
```

File: byoeb-v1/byoeb/byoeb/background_jobs/did_you_know/send_dyk.py (hunk 7 above)
Comment (mamuqsit) [reply byoeb-v1/byoeb/byoeb/background_jobs/did_you_know/send_dyk.py]: https://github.com/A4i-tech/byoeb/pull/34/commits/be2bc60bd04aaef4db494a01ac46d92711b0e7b7 addresses the disambiguation

File: byoeb-v1/byoeb/byoeb/background_jobs/did_you_know/send_dyk.py (hunk 5 above)
Comment (mamuqsit) [reply byoeb-v1/byoeb/byoeb/background_jobs/did_you_know/send_dyk.py]: Addressed by https://github.com/A4i-tech/byoeb/pull/34/commits/c334f6c0d3be80eacaf9862f5c0532fb4ccbf90e

File: byoeb-v1/byoeb/byoeb/background_jobs/did_you_know/send_dyk.py (hunk 4 above)
Comment (mamuqsit) [reply byoeb-v1/byoeb/byoeb/background_jobs/did_you_know/send_dyk.py]: Addressed by https://github.com/A4i-tech/byoeb/pull/34/commits/c9b01bda6847c1bd5874c79ec1ec65ebec56e022

**Comment by mamuqsit** [comment]
@Soumabha Up for review once again, I've introduced a DYK repository and batched the workflow against userbase.

Batching is handled by pick_candidates and is a core prerequisite of the other individual ops (which now ingest `DykBatch`). DB calls are read + update heavy, inserts are only done once per batch. All update calls do an `_id` lookup.

DB calls for one complete run with 1 batch: 6 reads, 3 updates, 1 insert.
For N batches: 3 + 3N reads, 2N + 1 updates, N inserts.

A failed batch costs 2 read calls and 2 update calls for a retry.

File: byoeb-v1/byoeb/byoeb/background_jobs/dependency_setup.py [hunk 10]
```diff
@@ -88,4 +88,16 @@
     api_key=env_config.env_openai_api_key,
     api_version=app_config["llms"]["openai"]["api_version"],
     organization=env_config.env_openai_org_id
-)
\ No newline at end of file
+)
+
+from byoeb.application_logger.azure_app_insights import AzureAppInsightsLogger
```
Comment (Soumabha) [inline byoeb-v1/byoeb/byoeb/background_jobs/dependency_setup.py:93]: Are we going to use the config of background_jobs or the chat_app now that we enable scheduler runs with that. In case there is a reason for both to exist, we should possibly add single responsibilty comments and explain the difference.

File: byoeb-v1/byoeb/byoeb/repositories/user_repository.py [hunk 11]
```diff
@@ -32,6 +32,11 @@ async def find_users_by_district(self, district: str) -> List[Dict[str, Any]]:
         """Find users by district."""
         pass
 
+    @abstractmethod
+    async def find_test_users(self) -> List[Dict[str, Any]]:
```
Comment (Soumabha) [inline byoeb-v1/byoeb/byoeb/repositories/user_repository.py:36]: We should instead have a function which accepts a collection of user types and then returns user data as per that.

File: byoeb-v1/byoeb/byoeb/repositories/dyk_repository.py [hunk 12]
```diff
@@ -0,0 +1,99 @@
+from abc import ABC, abstractmethod
+from typing import Dict, Iterable, List, Set
+
+from byoeb.constants.user_enums import LanguageCode
+from byoeb.models.dyk import DykRecord
+
+class DykRepository(ABC):
```
Comment (Soumabha) [inline byoeb-v1/byoeb/byoeb/repositories/dyk_repository.py:7]: Please check the files within byoeb > services > databases > mongo_db > user_db.py . As per the current structure we are using these *_db.py as the respective repository classes and then injecting them in the respective `dependency_setup` files.

Probably we should create a dyk_db.py to emulate the dyk storage for users.

File: byoeb-v1/byoeb/byoeb/background_jobs/did_you_know/send_dyk.py [hunk 13]
```diff
@@ -0,0 +1,317 @@
...
+import json
+import logging
+import random
+import uuid
+import sys
+from byoeb.background_jobs.did_you_know.config import bot_config, current_dir
+from byoeb.background_jobs.dependency_setup import app_insights_logger, channel_client_factory
+from byoeb.models.dyk import DykFactSheet, DykRecord
+from byoeb.repositories.dyk_repository import DykRepository
+from byoeb.repositories.user_repository import UserRepository
```
Comment (Soumabha) [inline byoeb-v1/byoeb/byoeb/background_jobs/did_you_know/send_dyk.py:12]: Lets use the user_db.py which is the repository as per the current structure.

File: byoeb-v1/byoeb/byoeb/background_jobs/dependency_setup.py (hunk 10 above)
Comment (mamuqsit) [reply byoeb-v1/byoeb/byoeb/background_jobs/dependency_setup.py:93]: Both background_jobs files need not coexist - one is a duplicate (or at best a lite version of the other). In prior setup, these scripts were intended to run standalone (without requiring a BYOEB instance) due to crontab limitations.

After moving to APScheduler, these files (`dependency_setup.py`, `app_config.py`) have become boilerplate (there is no longer a need for them to be standalone - and infact these 2 files just add more project maintenance overhead). I propose deferring their cleanup to a separate follow-up task after merging #34 (this) and #41 (APScheduler) - thoughts?

File: byoeb-v1/byoeb/byoeb/repositories/user_repository.py (hunk 11 above)
Comment (mamuqsit) [reply byoeb-v1/byoeb/byoeb/repositories/user_repository.py:36]: Sounds good. I am guessing a function like that should also return test users for an empty list, right?
```py
await repo.find_users_by_types([])  # test users
await repo.find_users_by_types(["asha"])  # only asha
```

File: byoeb-v1/byoeb/byoeb/repositories/dyk_repository.py (hunk 12 above)
Comment (mamuqsit) [reply byoeb-v1/byoeb/byoeb/repositories/dyk_repository.py:7]: ~~Hm I am conflicted on responsibilities of repository and service classes - I notice UserDB returns high-level User objects, and so does its respective Repository implementatiton.~~

~~If both (Repository and Service) expose the same features, we are effectively writing more code to express one-and-the-same functionality, right?~~

File: byoeb-v1/byoeb/byoeb/repositories/user_repository.py (hunk 11 above)
Comment (mamuqsit) [reply byoeb-v1/byoeb/byoeb/repositories/user_repository.py:36]: Alternatively, we can make the function more powerful:
```py
from typing import List, Literal, TypeAlias, Union

TUserType: TypeAlias = Literal[*(e.value for e in UserType), "test"]  # ['asha', 'anm', 'others', 'test']
def find_users_by_types(types: Union[TUserType, List[TUserType]]):
    pass
```

Test users need special handling because `test_user` is stored alongside `user_type`:
```js
{
  User: {
    user_id: 'b63b2c24f5024230d6fa6247xxxxxxxx',
    user_name: 'Name User',
    user_location: {},
    user_language: 'en',
    user_type: 'asha',
    phone_number_id: '91xxxxxxxxxx',
    test_user: true,
    experts: {},
    audience: [],
    created_timestamp: 1761037035,
    activity_timestamp: null,
    last_conversations: [],
    additional_info: {}
  }
}
```

**Comment by Soumabha** [comment]
Closing in favour of #46
